content_api/new_upload/readers/newgrf.py

```diff
@@ -0,0 +1,464 @@
+import enum
+import hashlib
+
+from io import BytesIO
+
+from ...helpers.enums import PackageType
+from ..exceptions import ValidationException
+from .helpers import binreader
+
+CTRL_CODES = {
+    0x01: (1, " "),  # SETX
+    0x0E: (0, ""),  # SMALLFONT
+    0x0F: (0, ""),  # BIGFONT
+    0x1F: (2, " "),  # SETXY
+    0x7B: (0, " "),
+    0x7C: (0, " "),
+    0x7D: (0, " "),
+    0x7E: (0, " "),
+    0x7F: (0, " "),
+    0x80: (0, " "),
+    0x81: (2, " "),
+    0x82: (0, " "),
+    0x83: (0, " "),
+    0x84: (0, " "),
+    0x85: (0, " "),
+    0x86: (0, " "),
+    0x87: (0, " "),
+    0x88: (0, ""),  # BLUE
+    0x89: (0, ""),  # SILVER
+    0x8A: (0, ""),  # GOLD
+    0x8B: (0, ""),  # RED
+    0x8C: (0, ""),  # PURPLE
+    0x8D: (0, ""),  # LTBROWN
+    0x8E: (0, ""),  # ORANGE
+    0x8F: (0, ""),  # GREEN
+    0x90: (0, ""),  # YELLOW
+    0x91: (0, ""),  # DKGREEN
+    0x92: (0, ""),  # CREAM
+    0x93: (0, ""),  # BROWN
+    0x94: (0, ""),  # WHITE
+    0x95: (0, ""),  # LTBLUE
+    0x96: (0, ""),  # GRAY
+    0x97: (0, ""),  # DKBLUE
+    0x98: (0, ""),  # BLACK
+    0x99: (0, ""),  # BLUE
+    0x9E: (0, "\u20AC"),
+    0x9F: (0, "\u0178"),
+    0xA0: (0, "\u25B2"),
+    0xAA: (0, "\u25BC"),
+    0xAC: (0, "\u2713"),
+    0xAD: (0, "\u274C"),
+    0xAF: (0, "\u25B6"),
+    0xB4: (0, ""),  # Train
+    0xB5: (0, ""),  # Lorry
+    0xB6: (0, ""),  # Bus
+    0xB7: (0, ""),  # Plane
+    0xB8: (0, ""),  # Ship
+    0xB9: (0, "\u208B\u2081"),
+    0xBC: (0, "\u2191"),
+    0xBD: (0, "\u2193"),
+}
+
+EXT_CTRL_CODES = {
+    0x01: (0, " "),
```

Comment:
similar here - why some have a space and some not?

Reply:
Some control codes resemble special whitespace or characters, while some resemble formatting (colour, font size).